### PR TITLE
GitHub Action to run the tests on every pull request

### PR DESCRIPTION
**Pull request by cclauss**
Test output: https://github.com/cclauss/maxdevtools/actions

### CODE REVIEW

**Comment by cclauss** [comment]
Rebased to the `main` branch just to prove that #19 fixed the tests.
#19

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+      matrix:
+        python-version: [3.x]
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-python@v5
+        with:
+          python-version: ${{ matrix.python-version }}
+      - run: |
+         cd maxdiff
```
Comment (MattijsKneppers) [inline .github/workflows/ci.yml:19]: I wonder why `test.py` is ran here, and then after pytest is installed later, it is ran again?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+name: test
+on:
+  push:
```
Comment (MattijsKneppers) [inline .github/workflows/ci.yml:3]: Would you agree that it is not necessary to run the tests for every Push action, but only as a required check before merging?

**Review by MattijsKneppers — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 2 above)
Comment (jamesb93) [reply .github/workflows/ci.yml:3]: I was actually looking to see if anyone had already done something like this for PRs. Would be most handy. I agree checking every push is unnecessary right now, and just making sure tests pass for each PR would be ideal.

Comment (cclauss) [reply .github/workflows/ci.yml:3]: GitHub's Python docs seem to favor `on: [push]` but I am flexible.  It is useful if contributors and reviewers can see the tests pass before starting the code review.
* https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-python#using-multiple-python-versions

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        python-version: [3.x]
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-python@v5
+        with:
+          python-version: ${{ matrix.python-version }}
+      - run: |
+         cd maxdiff
+         python3 tests/test.py
```
Comment (cclauss) [inline .github/workflows/ci.yml:20]:
```suggestion
```

File: .github/workflows/ci.yml (hunk 1 above)
Comment (cclauss) [reply .github/workflows/ci.yml]: Removed.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (MattijsKneppers) [reply .github/workflows/ci.yml:3]: Thanks @cclauss. I am not very familiar with setting up github checks, so do I understand correctly that having them on push would re-trigger the checks on every push? In that case I agree having them on push makes sense.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (MattijsKneppers) [reply .github/workflows/ci.yml]: Thanks!

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+name: test
+on:
+  push:
+  pull_request:
+  workflow_dispatch:
```
Comment (cclauss) [inline .github/workflows/ci.yml:5]:
```suggestion
on: [push]
```

File: .github/workflows/ci.yml (hunk 2 above)
Comment (cclauss) [reply .github/workflows/ci.yml]: Simplified to just `on: [push]`.

**Comment by cclauss** [comment]
If you are going to run `prbase` then you should probably enable contributors to rebase their PRs on the current branch.
* https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-rebasing-for-pull-requests

**Comment by MattijsKneppers** [comment]
@cclauss thanks for all the updates. I requested for rebasing to be enabled and will let you know when that has happened.

In the mean time, I wonder if you have noticed the requirement in the [contributing guidelines](https://github.com/Ableton/maxdevtools/blob/main/maxdiff/CONTRIBUTING.md) that Ableton has received a CLA from any contributors. Would you be able to submit one for us?

**Comment by cclauss** [comment]
Done.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (cclauss) [reply .github/workflows/ci.yml]: The nice explanation can be found at
* FactoryBoy/factory_boy#1075

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,17 @@
+name: test
+on: [push]
```
Comment (ala-ableton) [inline .github/workflows/ci.yml:2]: This means the workflow will trigger when a branch in this repository (not in a fork) is pushed to. So when an external contributor opens a new pull request, it won't trigger the workflow.

Comment (cclauss) [reply .github/workflows/ci.yml]: Changed to follow the factory-boy example.

**Review by MattijsKneppers — CHANGES_REQUESTED** [review]
Hi @cclauss, we had another look and have a few questions:
- it looks like `pull_request` needs to be `pull_request_target` in order to run from forks, would you be able to change that?
- nitpick: 
    - since we're only testing a single python version, perhaps having the `matrix` entry is not needed and the `3.x` version can just be in-lined in `python-version: ${{ matrix.python-version }}`.
    - OR: since `ubuntu-latest` already comes with python 3.10.x, perhaps `actions/setup-python@v5` is not needed at all.

Thanks again for your ongoing attention to this!

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  push:
+    branches:
+      - "main"
+  pull_request:
+jobs:
+  test:
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: [3.x]
```
Comment (cclauss) [inline .github/workflows/ci.yml:12]:
```suggestion
```

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-python@v5
+        with:
+          python-version: ${{ matrix.python-version }}
```
Comment (cclauss) [inline .github/workflows/ci.yml:18]:
```suggestion
          python-version: “3.10”
```

**Comment by cclauss** [comment]
Do these three suggestions cover it?

**Comment by MattijsKneppers** [comment]
@cclauss they do, thank you!

**Review by MattijsKneppers — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+      - "main"
+  pull_request_target:
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-python@v5
+        with:
+          python-version: “3.10”
```
Comment (ala-ableton) [inline .github/workflows/ci.yml:14]: I don't think these quotes will be accepted.

Comment (cclauss) [reply .github/workflows/ci.yml:14]: The quotes are VITAL in YAML.  This is the Yaml Norway problem.

https://dev.to/hugovk/the-python-3-1-problem-85g

Comment (cclauss) [inline .github/workflows/ci.yml:14]:
```suggestion
          python-version: "3.10"
```

**Comment by MattijsKneppers** [comment]
Hi @cclauss, sorry for the long wait. We needed a bit more discussion, but it's all approved now. It would be great if you can rebase one last time and then I can merge.

**Comment by cclauss** [comment]
Rebased.

**Review by ala-ableton — CHANGES_REQUESTED** [review]
The current state of `.github/workflows/ci.yml` is an old state, and is not what we want to merge to this repository.

**Pull request merged**